polkadot/node/network/statement-distribution/src/v2/requests.rs

```diff
@@ -422,28 +426,37 @@ impl RequestManager {
 /// A manager for pending responses.
 pub struct ResponseManager {
 	pending_responses: FuturesUnordered<BoxFuture<'static, TaggedResponse>>,
+	active_peers: HashSet<PeerId>,
 }
 
 impl ResponseManager {
 	pub fn new() -> Self {
-		Self { pending_responses: FuturesUnordered::new() }
+		Self { pending_responses: FuturesUnordered::new(), active_peers: HashSet::new() }
 	}
 
 	/// Await the next incoming response to a sent request, or immediately
 	/// return `None` if there are no pending responses.
 	pub async fn incoming(&mut self) -> Option<UnhandledResponse> {
-		self.pending_responses
-			.next()
-			.await
-			.map(|response| UnhandledResponse { response })
+		if let Some(response) = self.pending_responses.next().await {
+			// Upon receiving a response, remove the peer from the active set.
+			self.active_peers.remove(&response.requested_peer);
+			return Some(UnhandledResponse { response });
+		}
+		None
```

Comment:
This looks it can be written like prev impl:
```suggestion
                self.pending_responses.next().await.map(|response| {
                          self.active_peers.remove(&response.requested_peer);
                          UnhandledResponse { response }
                })
```